grails-app/controllers/org/pih/warehouse/notification/NotificationController.groovy

```diff
@@ -104,63 +90,206 @@ class NotificationController {
     def createProduct = {
         JSONObject json = request.getJSON() as JSONObject
         log.info "json for create Product::${json}"
-        if (json.has("Type") && json.getString("Type") == "SubscriptionConfirmation") {
-            String token = json.getString("Token")
-            try {
-                String PRODUCT_TOPIC_ARN = grailsApplication.config.awssdk.sns.product.arn
-                log.info "PRODUCT_TOPIC_ARN::${PRODUCT_TOPIC_ARN}, confirmation Token:${token}"
-                ConfirmSubscriptionRequest request = new ConfirmSubscriptionRequest(PRODUCT_TOPIC_ARN, token)
-                ConfirmSubscriptionResult result = amazonSnsClient.confirmSubscription(request);
-                log.info "result::${result?.toString()}"
-            } catch (Exception e) {
-                System.err.println(e.printStackTrace());
-            }
+
+        if (notificationService.confirmSubscription(json, Constants.ARN_TOPIC_PRODUCT_CREATE_UPDATE_TYPE)){
             render "subscribed"
             return
         }
-
         String message = json.getString("Message")
         JSONObject productsJson = new JSONObject(message)
         JSONArray products = productsJson.getJSONArray("products")
         for (int i = 0; i < products.length(); i++) {
             JSONObject pJson = products.getJSONObject(i)
-            log.info "PRODUCT JSON ${pJson} at index:${pJson}"
-            Category category = Category.findById(pJson.getString("category"))
-            if (!category) {
-                category = new Category()
-                category.id = pJson.getString("category")
-                category.name = pJson.getString("category")
-                category.save()
+            Product productInstance = productService.createFromJson(pJson)
+        }
+    }
+
+    def subscribeOrder = {
+        String ORDER_TOPIC_ARN = grailsApplication.config.awssdk.sns.order.arn?.toString()
+
+        log.info "Subscribiing topic:${ORDER_TOPIC_ARN}"
+        try {
+            String subscribeUrl = g.createLink(uri: "/api/notifications/orders", absolute: true)
+            if (grailsApplication.config.openboxes.integration.sns.orders.subscribeUrl) {
+                subscribeUrl = grailsApplication.config.openboxes.integration.sns.orders.subscribeUrl
+            }
+            log.info "Order subscribeUrl::${subscribeUrl}"
+            SubscribeRequest subscribeRequest = new SubscribeRequest(ORDER_TOPIC_ARN, "https", subscribeUrl)
+            subscribeRequest.returnSubscriptionArn = true
+            SubscribeResult result = notificationService.amazonSnsClient.subscribe(subscribeRequest);
+            log.info("Subscription ARN is " + result.subscriptionArn);
+        } catch (Exception e) {
+            log.error(e.printStackTrace());
+        }
+        render "subscribed"
+    }
+
+    def createOrder = {
```

Comment:
Let's move the two main methods into their own controllers. 

- ProductNotificationController
- OrderNotificationController 

and change the action to handleMessage(). The URL Mappings can still be whatever we called them before but just point them to the new controllers
```
/api/notifications/products -> ProductNotificationController.handleMessage()
/api/notifications/orders -> OrderNotificationController.handleMessage()
```